File: 2020/3-5-Python-Request-https-with-proxy/verify_https.py
# -*- encoding: utf-8 -*-
'''
@File    :   verify_https.py
@Time    :   2020/03/05 16:23:06
@Author  :   play4fun
@Desc    :   https请求安全性，验证证书
'''

import requests
def main():
    
    url='https://httpbin.org/ip'
    proxy='http://127.0.0.1:8080'
    proxies = {
        'http': proxy,
        'https': proxy,
    }
    try:
        rs = requests.get(url, proxies=proxies)
        print(rs.text)
        #报错 certificate verify failed
    except Exception as e:
        print(e)
    print('-'*30)

    rs = requests.get(url)
    print('正常:',rs.text)
    #正常
    print('-'*30)

    #取消证书验证
    rs = requests.get(url, proxies=proxies,verify=False)
    print(rs.text)
    pass

if __name__ == "__main__":
    main()
    pass

File: 2020/3-5-Python-Request-https-with-proxy/verify_https.go
/*
@File    :   verify_https.go
@Time    :   2020/03/07 09:38:20
@Author  :   play4fun
@Desc    :   同样的逻辑，用Go语言实现
*/

package main

import (
	"crypto/tls"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"net/url"
)

func main() {
	url2 := "https://httpbin.org/ip" // 小心同名 "net/url"
	proxy := "http://127.0.0.1:8080" //mitmproxy
	// proxy := "http://42.56.89.102:6159" //代理服务器Squid,OK

	urli := url.URL{}
	urlproxy, _ := urli.Parse(proxy)
	c := http.Client{
		Transport: &http.Transport{
			Proxy:           http.ProxyURL(urlproxy),               //设置代理
			TLSClientConfig: &tls.Config{InsecureSkipVerify: true}, //跳过证书验证
		},
	}
	if resp, err := c.Get(url2); err != nil {
		log.Fatalln(err)
		//2020/03/07 09:45:58 Get https://httpbin.org/ip: x509: certificate signed by unknown authority
		//exit status 1
	} else {
		defer resp.Body.Close()
		body, _ := ioutil.ReadAll(resp.Body)
		fmt.Printf("%s\n", body)
	}
}
